src/rust.rs

```diff
@@ -270,41 +268,103 @@ impl RootKind for Value {
     fn rootKind() -> jsapi::RootKind { jsapi::RootKind::Value }
 }
 
-impl<T: RootKind + Copy> Rooted<T> {
-    pub fn new_with_addr(cx: *mut JSContext, initial: T, addr: *const u8) -> Rooted<T> {
-        let ctxfriend: &mut ContextFriendFields = unsafe {
-            mem::transmute(cx)
-        };
-
-        let kind = T::rootKind() as usize;
-        let root = Rooted::<T> {
+impl<T> Rooted<T> {
+    pub fn new_unrooted(initial: T) -> Rooted<T> {
```

Comment:
That trait seems like a good choice to have as a bound in the struct itself.
Alright, other changes suggested on IRC were to have `struct Rootable<T>(Rooted<T>);` which could only be created empty and not accessed, and it's what `RootedGuard` would accept.